org_settings: Add feature to change bot role from UI.

somesh202 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
The first commit: 
- Adds a new column in the bot-list table in the organization settings to show the role a bot has been assigned.

The second commit: 

- Adds the `Role` field in the `Manage bot` modal, allowing an admin to change roles of a bot through the UI and fixing the 
  formatting and ordering of fields in the modal.

The third commit: 
- Adds the main feature to change the roles of a bot through the UI. 

Fixes: #21105

**Testing plan:** <!-- How have you tested? -->
Manually
**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![Screenshot 2022-02-25 171455](https://user-images.githubusercontent.com/77766761/155833545-78f1c102-7064-40fa-a4a0-fd35e1d57f77.png)
![Screenshot 2022-03-01 204419](https://user-images.githubusercontent.com/77766761/156195419-78a14eee-dcec-422a-9ae3-295b0dc56421.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on static/js/settings_users.js:
```diff
@@ -625,16 +626,21 @@ function handle_bot_form(tbody, status_field) {
             const full_name = $("#dialog_widget_modal").find("input[name='full_name']");
 
             const url = "/json/bots/" + encodeURIComponent(user_id);
+            const url_for_role_change = "/json/users/" + encodeURIComponent(user_id);
```
@timabbott Do we want to use `/json/users` url for role or `/json/bots`?

sahil839 commented on static/js/settings_users.js:
```diff
@@ -667,14 +674,19 @@ function handle_bot_form(tbody, status_field) {
+            $("#bot-role-select").val(bot.role);
+            $("#bot-role-select")
+                .find(`option[value="${CSS.escape(settings_config.user_role_values.owner.code)}"]`)
+                .hide();
```
Have we decided to not allow a bot with role `Owner`? I don't think it is mentioned in the issue but we can confirm. And if so, we should make sure we add the check in API too.

somesh202 replied:
It wasn't mentioned in the issue, but I just thought why will a bot get the `Owner` role. 
If that's not the case, shall I allow the `Owner` role for bots?

sahil839 replied:
I am not sure too, but for various purposes we check the permission of its owner. @alya thoughts?

sahil839 commented on static/templates/settings/admin_bot_form.hbs:
```diff
@@ -1,20 +1,39 @@
+        <div class="email_change_container">
+            <label for="email">{{t "Email" }}</label>
+            <input type="text" autocomplete="off" name="email" value="{{ email }}" readonly/>
+        </div>
```
The spacing between the inputs is inconsistent. Space is less between name and email and between email and ID as compared to others.

somesh202 replied:
Done. I've updated the modal screenshot in the PR description.

sahil839 commented on static/templates/settings/admin_user_list.hbs:
```diff
@@ -12,14 +12,16 @@
     </td>
-    {{#unless is_bot}}
+    {{/unless}}
+
```
We should avoid adding these unnecessary lines.

sahil839 commented on static/js/settings_users.js:
```diff
@@ -58,7 +58,9 @@ function sort_bot_email(a, b) {
-    return compare_a_b(a.role, b.role);
+    const role_a = people.get_user_type(a.user_id) || "";
+    const role_b = people.get_user_type(b.user_id) || "";
+    return compare_a_b(role_a.toLowerCase(), role_b.toLowerCase());
```
We want to sort the role column according to hierarchy of role not alphabetically. So, there is no need to change this.

sahil839 reviewed (commented):
Thanks for working on this @somesh202. I left some comments.

Also you can split these into multiple commits probably with each commit being a single point in the issue wherever possible.

timabbott commented:
@somesh202 is this ready for a next review?  It'd be helpful if you posted a comment summarizing how you addressed all the feedback, if so.

somesh202 commented:
@timabbott  Yes, I have made the changes suggested by @sahil839.

- Fixed the inconsistency of spaces between some fields in the modal by adding an `input_group` class that takes care of the CSS.
- Reverted the change in the sorting of the `Role` column which I misunderstood to be alphabetically.
- I'm not sure about the fact that whether we want to allow the bot to have an `Owner` role, which we didn't had in the case of a normal user.

sahil839 commented:
> * I'm not sure about the fact that whether we want to allow the bot to have an `Owner` role, which we didn't had in the case of a normal user.

We have owner role option for normal users, only that only an existing owner can see it. You can login as `Desdemona` in dev server to check.

somesh202 commented:
@sahil839 Pushed some changes, which would allow an Owner to give an owner role to a bot (as in the case of normal users).

sahil839 commented on static/js/settings_users.js:
```diff
@@ -684,14 +692,25 @@ function handle_bot_form(tbody, status_field) {
 
         dialog_widget.launch({
-            html_heading: $t_html({defaultMessage: "Change bot info and owner"}),
+            html_heading: $t_html({defaultMessage: "Manage bot"}),
```
This should be in the previous commit as per the commit message.

sahil839 commented on static/js/settings_users.js:
```diff
@@ -196,6 +196,7 @@ function bot_info(bot_user_id) {
     info.user_id = bot_user.user_id;
     info.full_name = bot_user.full_name;
     info.bot_owner_id = owner_id;
+    info.user_role_text = people.get_user_type(bot_user_id);
```
This should be in the first commit where you are adding role column.

sahil839 commented on static/js/settings_users.js:
```diff
@@ -281,7 +282,7 @@ section.bots.create_table = () => {
             email: sort_bot_email,
             bot_owner: sort_bot_owner,
-            id: sort_user_id,
+            role: sort_role,
```
This should be in the first commit where you are adding role column.

sahil839 reviewed (commented):
@somesh202 thanks for updating the PR. I found a couple of instances where the change is in the wrong commit which should be fixed. This prevents us from merging commits incrementally and also may confuse someone who is going through git commit history. Also, I would encourage you to check all the commits once and fix if there are other changes like these. Rest the PR looks good, apart from the url part that Tim will provide feedback on.

somesh202 commented:
@sahil839 I've updated the PR by adding the changes to their respective commits.

sahil839 commented:
I posted some comments above. Also would be nice to move the commit adding the `role change` feature to last if we need to decide about what url to use and then we can atleast merge the other commits which improves the UI first.

somesh202 commented on static/templates/settings/admin_bot_form.hbs:
```diff
@@ -1,19 +1,38 @@
+                {{t "Bot type" }}
+                <i class="fa fa-question-circle settings-info-icon bot_type_tooltip tippy-zulip-tooltip" aria-hidden="true" data-tippy-content='{{t "Incoming webhooks can only send messages." }}'></i>
+            </label>
+            <input type="text" autocomplete="off" name="bot_type" value="{{ bot_type }}" readonly/>
```
I added this field just to match the table contents, which was mentioned in the issue. Shall I remove this, if it's not that relevant?

sahil839 replied:
I am not sure, we can keep it for now since you have added it and remove it if required after feedback from Alya or Tim.

somesh202 commented:
@sahil839  I've updated the PR by shifting the `role change` commit at the last and `Adding the role field` change  to the last commit.

sahil839 commented on static/js/settings_users.js:
```diff
@@ -641,16 +641,22 @@ function handle_bot_form($tbody, $status_field) {
             full_name: bot.full_name,
+            user_role_values: settings_config.user_role_values,
+            disable_role_dropdown: bot.is_owner && !page_params.is_owner,
+            bot_type: settings_bots.type_id_to_string(bot.bot_type),
```
This bot type field should be added in previous commit.

sahil839 reviewed (commented):
Posted one comment. Also please update the commit messages as per the changes that you have done in that commit.

somesh202 commented:
@sahil839 Updated the PR and the commit messages as per your feedback.

sahil839 commented:
@somesh202 Is this PR closed by mistake?

somesh202 commented:
@sahil839 Yes, I don't know how it happened, all my open PRs got closed? Is there a solution to re-open this?

sahil839 commented:
I re-opened it.

sahil839 commented on static/js/settings_users.js:
```diff
@@ -641,17 +641,22 @@ function handle_bot_form($tbody, $status_field) {
+            const url_for_role_change = "/json/users/" + encodeURIComponent(user_id);
             const data = {
                 full_name: $full_name.val(),
+                role: JSON.stringify(role),
```
We should pass `role` field only to the `/json/users` url if we have decided to use this url only not to `/json/bots` url.

sahil839 reviewed (commented):
I posted one comment and you should remove `This also adds the new "Role" dropdown field to the modal` from the commit message of second commit as that change is done in third commit. Otherwise LGTM.

somesh202 commented:
@sahil839 I've created a new PR for this (#21627) addressing your feedback as the branches of this PR got deleted.

somesh202 commented:
Closing this in favour of #21627.
